### PR TITLE
[RFR] Translator core library

### PR DESCRIPTION
**Pull request by mickaelandrieu**
Translator component is OK, I will improve it.
After thinking, I think the best place for this module is to be part of core *because* it will be mandatory.

In all cases, component or core, the translator works: I will add more tests and probably try to implement pluralization system.

When I tried to aggregate the translator into Renderer (and finaly, into nunjucks), this won't work.
I investigate a lot (and I will suggest a "Debug" component in the future because Javascript is damn so hard to debug!), and *in fact* the **Renderer don't work as expected**.

- [x] Component translator
    - [x] Should return a value for a key in default catalog
    - [x] Should return a value for a key in another catalog when new locale is set
    - [x] Should be able to return default value if translation is not available in the selected catalog
    - [x] Should log a notice when this is not the translation from the selected catalog  
    - [x] Should return the key is the key is not available at all
    - [x] Should log a warning exception in this case
    - [x] Store all loaded catalogs into un object indexed by locale (*limit ajax calls by @ndufreche* )
- [x] Renderer integration
    - [x] Should register an event on application initialization
    - [x] Should make ``trans()`` fonction from translator available in Renderer in main App scope
    - [x] Should be tested
    - [x] Move logic of filters outside Renderer
- [x] Docs
    - [x] Translator
    - [x] Filters in Renderer

**update 20/02 17h49**: docs are to be done in another repo

**update 20/02 16h51**: the component work the way we expected :)

### CODE REVIEW

File: src/tb/core/Translator.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+         */
+    var Translator = new JS.Class({
+            initialize: function () {
+                this.config = {};
+                this.catalog = {};
+                /*var env = nunjucks.configure();
+                env.addFilter('i18n', this.translate, true);*/
+            },
+
+            init: function (config) {
```
Comment (fkroockmann) [inline src/tb/core/Translator.js]: You don't need have a second initializer, use initialize of JS class

File: src/tb/core/Translator.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                this.config = config;
```
Comment (fkroockmann) [inline src/tb/core/Translator.js]: You need to test if config is undefined or not

File: src/tb/core/Translator.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+            getDefaultLocale: function () {
+                return this.config.locale;
+            },
+
+            translate: function (key) {
+                if (this.catalog[key] !== undefined) {
+                    return this.catalog[key];
+                }
+                Core.exception('TranslatorException', 500, 'The key "' + key + '" is malformed.', {});
```
Comment (fkroockmann) [inline src/tb/core/Translator.js]: I think, if the key is undefined in the catalog you need to return something

File: src/tb/core/Translator.js (hunk 2 above)
Comment (mickaelandrieu) [inline src/tb/core/Translator.js]: if I set a config in config.json (on translator part), is this Ok to you ?

File: specs/tb/core/Renderer.spec.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+define(['tb.core.Renderer'], function (Renderer) {
+    'use strict';
+
+    describe("Renderer test suite", function () {
+        it("Should render simple strings", function () {
+            expect(Renderer.render('Hello world')).toEqual('Hello world');
+        });
+
+        it("Should render string with variables", function () {
+            expect(Renderer.render('Hello {{ world }}', {'world': 'world'})).toEqual('Hello world');
```
Comment (ndufreche) [inline specs/tb/core/Renderer.spec.js]: This is not a translation test

**Comment by fkroockmann** [comment]
Gruntfile must not be in this PR

File: src/tb/component/translator/main.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                this.locale = locale;
+            },
+
+            getDefaultLocale: function () {
+                return this.config.locale;
+            },
+
+            translate: function (key) {
+                if (this.catalog[key] !== undefined) {
+                    return this.catalog[key];
```
Comment (fkroockmann) [inline src/tb/component/translator/main.js]: Use and return a var

**Comment by mickaelandrieu** [comment]
Gruntfile ?

File: src/tb/component/translator/main.js (hunk 5 above)
Comment (mickaelandrieu) [inline src/tb/component/translator/main.js]: ok, but why ?

**Comment by fkroockmann** [comment]
Simple and effective :)

File: src/tb/component/translator/main.js (hunk 5 above)
Comment (fkroockmann) [inline src/tb/component/translator/main.js]: i'm not fan when using multi return.
The return of a function must be at the end ^^

File: src/tb/component/translator/main.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            getDefaultLocale: function () {
+                return this.config.locale;
+            },
+
+            translate: function (key) {
+                var translation = key;
+                if (this.catalog[key] !== undefined) {
+                    translation = this.catalog[key];
+                } else if (this.getDefaultCatalog()[key] !== undefined) {
+                    Logger.notice('The key "' + key + '" has not translation in the selected catalog.');
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: It is possible to set the log more explicit with the name of the catalog ?

File: src/tb/component/translator/main.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                    translation = this.getDefaultCatalog()[key];
+                } else {
+                    Logger.warning('The key "' + key + '" is malformed.');
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: "The key "' + key + '" is malformed or inexistant."

File: src/tb/component/translator/main.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            loadCatalog: function (locale) {
+                var self = this,
+                    myLocale = (locale === undefined) ? self.locale : locale;
+                jQuery.ajax({
+                    'url': self.base + '/' + myLocale + '/global.json',
+                    'data': 'json',
+                    'async': false
+                })
+                    .done(function (response) {
+                        self.catalog = JSON.parse(response, true);
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: JSON.parse(text[, reviver])
reviver Optional
If a function, prescribes how the value originally produced by parsing is transformed, before being returned.

You can remove the second attribute.

File: src/tb/component/translator/main.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                this.locale = config.locale;
+                this.loadCatalog();
+            },
+
+            getLocale:  function () {
+                return this.locale;
+            },
+
+            setLocale: function (locale) {
+                this.loadCatalog();
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: Your catalog is already loaded in the init function.

File: src/tb/component/translator/main.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+                return translation;
+            },
+
+            getCatalog: function () {
+                return this.catalog;
+            },
+
+            getDefaultCatalog: function () {
+                this.loadCatalog(this.getDefaultLocale());
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: Each time you don't find a value in the current catalog you download the default catalog ?

File: src/tb/component/translator/main.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+         * {
+         *      'hello_world' : 'Hello world'
+         * }
+         */
+    var Translator = {
+            init: function (config) {
+                this.base = config.base;
+                this.catalog = {};
+                this.config = config;
+                this.locale = config.locale;
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: you need a default local too.

File: src/tb/component/translator/main.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+            getDefaultLocale: function () {
+                return this.config.locale;
+            },
+
+            translate: function (key) {
+                var translation = key;
+                if (this.catalog[key] !== undefined) {
+                    translation = this.catalog[key];
+                } else if (this.getDefaultCatalog()[key] !== undefined) {
```
Comment (ndufreche) [inline src/tb/component/translator/main.js]: You have to be restrictive
```javascript
} else if (this.locale !== this.default_local && thisthis.getDefaultCatalog()[key] !== undefined) {
```

File: src/tb/component/translator/main.js (hunk 9 above)
Comment (mickaelandrieu) [inline src/tb/component/translator/main.js]: but i need to relaod it after load a new locale ;)

File: src/tb/component/translator/main.js (hunk 8 above)
Comment (fkroockmann) [inline src/tb/component/translator/main.js]: I think JSLint want it

File: specs/require.config.js [hunk 13]
```diff
@@ -4,11 +4,12 @@ require.config({
     paths: {
         'tb.core': 'src/tb/main',
         'component': 'src/tb/component/component',
+
         'jquery': 'bower_components/jquery/dist/jquery.min',
         'jqueryui': 'bower_components/jquery-ui/jquery-ui.min',
         'jsclass' : 'node_modules/jsclass/min/core',
         'underscore': 'bower_components/underscore/underscore-min',
-        'nunjucks': 'bower_components/nunjucks/browser/nunjucks.min',
```
Comment (fkroockmann) [inline specs/require.config.js]: let min :)

File: specs/tb/component/translator/Translator.spec.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
+define(['component!translator', 'component!logger', 'tb.core.Renderer', 'es5-shim/es5-shim'], function (Translator, Logger, Renderer) {
+    'use strict';
+
+    Translator.init({'locale': 'en_US', 'base': 'src/tb/i18n/'});
+    var locale_fr = "fr_FR";
+
+    describe("Translator core library test suite", function () {
+        it("Should translate the key from default catalog", function () {
```
Comment (fkroockmann) [inline specs/tb/component/translator/Translator.spec.js]: More spaces more smooth :)

**Comment by fkroockmann** [comment]
Good job :)

**Comment by mickaelandrieu** [comment]
@ndufreche this is all green, please merge :bear:

**Comment by ndufreche** [comment]
Nice job :+1:

**Pull request merged**